File: leetcode-78.py
from typing import List

class Solution:
    def subsets(self, nums: List[int]) -> List[List[int]]:
        ans = []
        lenq = len(nums)
        def dfs(u, li):
            if u == lenq:
                ans.append(li[:])
                return
            dfs(u+1, li+[nums[u]])
            dfs(u+1, li)
        dfs(0, [])
        return ans


if __name__ == "__main__":
    so = Solution()
    print(so.subsets([1,2,3]))

File: leetcode/leetcode-78.cpp
/*
 *@File Name          :leetcode-78.cpp
 *@Author             :cc
 *@Version            :1.0.0
 *@Date               :2022/7/8 9:57
 *@Description        :
 *@Function List      :
 *@History            :
 */

#include "algorithm"
#include "vector"
using namespace std;

class Solution {
 public:
  vector<vector<int>> ans;

  void dfs(vector<int>& nums, int cur, vector<int>& t) {
    if (cur == nums.size()) {
      ans.push_back(t);
      return;
    }
    dfs(nums, cur + 1, t);
    t.push_back(nums[cur]);
    dfs(nums, cur + 1, t);
    t.pop_back();
  }

  vector<vector<int>> subsets(vector<int>& nums) {
    std::sort(nums.begin(), nums.end());
    nums.erase(std::unique(nums.begin(), nums.end()), nums.end());
    vector<int> t;
    dfs(nums, 0, t);
    return ans;
  }
};